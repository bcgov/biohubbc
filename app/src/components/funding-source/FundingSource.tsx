import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import Typography from '@mui/material/Typography';
import { DATE_FORMAT } from 'constants/dateTimeFormats';
import { getFormattedAmount, getFormattedDateRangeString } from 'utils/Utils';

export interface IFundingSource {
  id: number;
  agency_name?: string;
  investment_action_category_name?: string;
  funding_amount?: number;
  start_date?: string;
  end_date?: string;
  agency_project_id: string;
  first_nations_name?: string;
}

interface IFundingSourceProps {
  funding_sources: IFundingSource[];
}

/**
 * Funding source content for a project.
 *
 * @return {*}
 */
const FundingSource = (props: IFundingSourceProps) => {
  const hasFundingSources = props.funding_sources.length > 0;
  return (
    <>
      <List disablePadding>
        {hasFundingSources &&
          props.funding_sources.map((item: IFundingSource) => (
            <ListItem disableGutters divider key={item.id}>
              <Box flex="1 1 auto">
                <Box pb={1.25}>
                  <Typography component="span">
                    {item.agency_name ?? item.first_nations_name}
                    {item.investment_action_category_name &&
                      item.investment_action_category_name !== 'Not Applicable' && (
                        <Typography component="span">&nbsp;({item.investment_action_category_name})</Typography>
                      )}
                  </Typography>
                </Box>
                <Box component="dl" m={0}>
                  <Grid container spacing={1}>
                    <Grid item sm={6}>
                      <Typography component="dt" variant="subtitle2" color="textSecondary">
                        Project ID
                      </Typography>
                      <Typography component="dd">{item.agency_project_id || 'No Agency Project ID'}</Typography>
                    </Grid>
                    <Grid item sm={6}>
                      {item.start_date && item.end_date && (
                        <>
                          <Typography component="dt" variant="subtitle2" color="textSecondary">
                            Timeline
                          </Typography>
                          <Typography component="dd">
                            {getFormattedDateRangeString(
                              DATE_FORMAT.ShortMediumDateFormat,
                              item.start_date,
                              item.end_date
                            )}
                          </Typography>
                        </>
                      )}
                    </Grid>
                    <Grid item sm={12}>
                      {item.funding_amount && (
                        <>
                          <Typography component="dt" variant="subtitle2" color="textSecondary">
                            Funding Amount
                          </Typography>
                          <Typography component="dd">{getFormattedAmount(item.funding_amount)}</Typography>
                        </>
                      )}
                    </Grid>
                  </Grid>
                </Box>
              </Box>
            </ListItem>
          ))}

        {!hasFundingSources && (
          <ListItem disableGutters>
            <Typography>No Funding Sources</Typography>
          </ListItem>
        )}
      </List>
    </>
  );
};

export default FundingSource;
