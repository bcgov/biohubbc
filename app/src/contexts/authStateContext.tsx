import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import React from 'react';

export interface IAuthState {
  ready?: boolean;
}

export const AuthStateContext = React.createContext<IAuthState>({
  ready: false
});

export const AuthStateContextProvider: React.FC = (props) => {
  const keycloakWrapper = useKeycloakWrapper();

  return (
    <AuthStateContext.Provider value={{ ready: keycloakWrapper.keycloak?.authenticated }}>
      {props.children}
    </AuthStateContext.Provider>
  );
};
