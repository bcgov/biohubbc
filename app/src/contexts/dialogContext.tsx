import IconButton from '@material-ui/core/IconButton';
import Snackbar from '@material-ui/core/Snackbar';
import CloseIcon from '@material-ui/icons/Close';
import { Color } from '@material-ui/lab/Alert';
import { ErrorDialog, IErrorDialogProps } from 'components/dialog/ErrorDialog';
import YesNoDialog, { IYesNoDialogProps } from 'components/dialog/YesNoDialog';
import React, { createContext, ReactNode, useState } from 'react';

export interface IDialogContext {
  /**
   * Set the yes no dialog props.
   *
   * Note: Any props that are not provided, will default to whatever value was previously set (or the default value)
   *
   * @memberof IDialogContext
   */
  setYesNoDialog: (props: Partial<IYesNoDialogProps>) => void;
  /**
   * The current yes no dialog props.
   *
   * @type {IYesNoDialogProps}
   * @memberof IDialogContext
   */
  yesNoDialogProps: IYesNoDialogProps;
  /**
   * Set the error dialog props.
   *
   * Note: Any props that are not provided, will default to whatever value was previously set (or the default value)
   *
   * @memberof IDialogContext
   */
  setErrorDialog: (props: Partial<IErrorDialogProps>) => void;
  /**
   * The current error dialog props.
   *
   * @type {IErrorDialogProps}
   * @memberof IDialogContext
   */
  errorDialogProps: IErrorDialogProps;
  /**
   * Set the snackbar props.
   *
   * Note: Any props that are not provided, will default to whatever value was previously set (or the default value)
   *
   * @memberof IDialogContext
   */
  setSnackbar: (props: Partial<ISnackbarProps>) => void;
  /**
   * The current snackbar props.
   *
   * @type {ISnackbarProps}
   * @memberof IDialogContext
   */
  snackbarProps: ISnackbarProps;
}

export interface ISnackbarProps {
  open: boolean;
  onClose: () => void;
  severity?: Color;
  color?: Color;
  snackbarMessage: ReactNode;
}

export const defaultYesNoDialogProps: IYesNoDialogProps = {
  dialogTitle: '',
  dialogText: '',
  open: false,
  onClose: () => {
    // default do nothing
  },
  onNo: () => {
    // default do nothing
  },
  onYes: () => {
    // default do nothing
  }
};

export const defaultErrorDialogProps: IErrorDialogProps = {
  dialogTitle: '',
  dialogText: '',
  open: false,
  onClose: () => {
    // default do nothing
  },
  onOk: () => {
    // default do nothing
  }
};

export const defaultSnackbarProps: ISnackbarProps = {
  snackbarMessage: '',
  open: false,
  onClose: () => {
    // default do nothing
  }
};

export const DialogContext = createContext<IDialogContext>({
  setYesNoDialog: () => {
    // default do nothing
  },
  yesNoDialogProps: defaultYesNoDialogProps,
  setErrorDialog: () => {
    // default do nothing
  },
  errorDialogProps: defaultErrorDialogProps,
  setSnackbar: () => {
    // default do nothing
  },
  snackbarProps: defaultSnackbarProps
});

/**
 * Wraps the provided children in a context that provides various modal dialogs/popups.
 *
 * @param {*} props
 * @return {*}
 */
export const DialogContextProvider: React.FC = (props) => {
  const [yesNoDialogProps, setYesNoDialogProps] = useState<IYesNoDialogProps>(defaultYesNoDialogProps);

  const [errorDialogProps, setErrorDialogProps] = useState<IErrorDialogProps>(defaultErrorDialogProps);

  const [snackbarProps, setSnackbarProps] = useState<ISnackbarProps>(defaultSnackbarProps);

  const setYesNoDialog = function (partialProps: Partial<IYesNoDialogProps>) {
    setYesNoDialogProps({ ...yesNoDialogProps, ...partialProps });
  };

  const setSnackbar = function (partialProps: Partial<ISnackbarProps>) {
    setSnackbarProps({ ...snackbarProps, ...partialProps });
  };

  const setErrorDialog = function (partialProps: Partial<IErrorDialogProps>) {
    setErrorDialogProps({ ...errorDialogProps, ...partialProps });
  };

  return (
    <DialogContext.Provider
      value={{
        setYesNoDialog,
        yesNoDialogProps,
        setErrorDialog,
        errorDialogProps,
        setSnackbar,
        snackbarProps
      }}>
      {props.children}
      <YesNoDialog {...yesNoDialogProps} />
      <ErrorDialog {...errorDialogProps} />
      <Snackbar
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center'
        }}
        open={snackbarProps.open}
        autoHideDuration={6000}
        onClose={() => setSnackbar({ open: false })}
        message={snackbarProps.snackbarMessage}
        action={
          <React.Fragment>
            <IconButton size="small" aria-label="close" color="inherit" onClick={() => setSnackbar({ open: false })}>
              <CloseIcon fontSize="small" />
            </IconButton>
          </React.Fragment>
        }
      />
    </DialogContext.Provider>
  );
};
