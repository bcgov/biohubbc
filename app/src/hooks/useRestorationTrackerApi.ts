import axios from 'axios';
import useAdminApi from './api/useAdminApi';
import useAxios from './api/useAxios';
import useCodesApi from './api/useCodesApi';
import useDraftApi from './api/useDraftApi';
import useExternalApi from './api/useExternalApi';
import useProjectApi, { usePublicProjectApi } from './api/useProjectApi';
import useSearchApi, { usePublicSearchApi } from './api/useSearchApi';
import useUserApi from './api/useUserApi';
import usePermitApi from './api/usePermitApi';
import { useContext } from 'react';
import { ConfigContext } from 'contexts/configContext';

/**
 * Returns a set of supported api methods.
 *
 * @return {*} object whose properties are supported api methods.
 */
export const useRestorationTrackerApi = () => {
  const config = useContext(ConfigContext);
  const apiAxios = useAxios(config?.API_HOST);

  const project = useProjectApi(apiAxios);

  const permit = usePermitApi(apiAxios);

  const search = useSearchApi(apiAxios);

  const codes = useCodesApi(apiAxios);

  const draft = useDraftApi(apiAxios);

  const user = useUserApi(apiAxios);

  const admin = useAdminApi(apiAxios);

  const external = useExternalApi(axios);

  const publicApis = {
    project: usePublicProjectApi(apiAxios),
    search: usePublicSearchApi(apiAxios)
  };

  return {
    project,
    permit,
    search,
    codes,
    draft,
    user,
    admin,
    external,
    public: publicApis
  };
};
