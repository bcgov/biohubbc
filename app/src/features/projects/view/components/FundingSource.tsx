import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import { Theme } from '@material-ui/core/styles/createMuiTheme';
import makeStyles from '@material-ui/core/styles/makeStyles';
import { mdiPencilOutline, mdiPlus, mdiTrashCanOutline } from '@mdi/js';
import Icon from '@mdi/react';
import EditDialog from 'components/dialog/EditDialog';
import { IErrorDialogProps } from 'components/dialog/ErrorDialog';
import { IYesNoDialogProps } from 'components/dialog/YesNoDialog';
import { H3ButtonToolbar } from 'components/toolbar/ActionToolbars';
import { DATE_FORMAT } from 'constants/dateTimeFormats';
import { AddFundingI18N, DeleteProjectFundingI18N, EditFundingI18N } from 'constants/i18n';
import { DialogContext } from 'contexts/dialogContext';
import ProjectFundingItemForm, {
  IProjectFundingFormArrayItem,
  ProjectFundingFormArrayItemInitialValues,
  ProjectFundingFormArrayItemYupSchema
} from 'features/projects/components/ProjectFundingItemForm';
import { APIError } from 'hooks/api/useAxios';
import { useRestorationTrackerApi } from 'hooks/useRestorationTrackerApi';
import { IGetAllCodeSetsResponse } from 'interfaces/useCodesApi.interface';
import { IGetProjectForViewResponse } from 'interfaces/useProjectApi.interface';
import React, { useContext, useState } from 'react';
import { getFormattedAmount, getFormattedDate, getFormattedDateRangeString } from 'utils/Utils';

const useStyles = makeStyles((theme: Theme) => ({
  fundingSourceTable: {
    '& .MuiTableCell-root': {
      verticalAlign: 'middle'
    }
  }
}));

export interface IProjectFundingProps {
  projectForViewData: IGetProjectForViewResponse;
  codes: IGetAllCodeSetsResponse;
  refresh: () => void;
}

/**
 * Funding source content for a project.
 *
 * @return {*}
 */
const FundingSource: React.FC<IProjectFundingProps> = (props) => {
  const classes = useStyles();

  const {
    projectForViewData: { funding, id },
    codes
  } = props;

  const restorationTrackerApi = useRestorationTrackerApi();

  const dialogContext = useContext(DialogContext);

  const defaultErrorDialogProps = {
    dialogTitle: EditFundingI18N.editErrorTitle,
    dialogText: EditFundingI18N.editErrorText,
    open: false,
    onClose: () => {
      dialogContext.setErrorDialog({ open: false });
    },
    onOk: () => {
      dialogContext.setErrorDialog({ open: false });
    }
  };

  const showErrorDialog = (textDialogProps?: Partial<IErrorDialogProps>) => {
    dialogContext.setErrorDialog({ ...defaultErrorDialogProps, ...textDialogProps, open: true });
  };

  const defaultYesNoDialogProps = {
    dialogTitle: DeleteProjectFundingI18N.deleteTitle,
    dialogText: DeleteProjectFundingI18N.deleteText,
    open: false,
    onClose: () => dialogContext.setYesNoDialog({ open: false }),
    onNo: () => dialogContext.setYesNoDialog({ open: false }),
    onYes: () => handleDeleteDialogYes()
  };

  const showYesNoDialog = (yesNoDialogProps?: Partial<IYesNoDialogProps>) => {
    dialogContext.setYesNoDialog({ ...defaultYesNoDialogProps, ...yesNoDialogProps });
  };

  const [fundingFormData, setFundingFormData] = useState({
    index: 0,
    values: ProjectFundingFormArrayItemInitialValues
  });

  const [openEditDialog, setOpenEditDialog] = useState(false);

  const handleDialogEditOpen = async (itemIndex: number) => {
    let fundingSourceValues: IProjectFundingFormArrayItem;

    if (itemIndex < funding.fundingSources.length) {
      // edit an existing funding source
      const fundingSource = funding.fundingSources[itemIndex];

      fundingSourceValues = {
        id: fundingSource.id,
        agency_id: fundingSource.agency_id,
        investment_action_category: fundingSource.investment_action_category,
        investment_action_category_name: fundingSource.investment_action_category_name,
        agency_project_id: fundingSource.agency_project_id,
        funding_amount: fundingSource.funding_amount,
        start_date: getFormattedDate(DATE_FORMAT.ShortDateFormat, fundingSource.start_date),
        end_date: getFormattedDate(DATE_FORMAT.ShortDateFormat, fundingSource.end_date),
        revision_count: fundingSource.revision_count
      };
    } else {
      // add a new funding source
      fundingSourceValues = ProjectFundingFormArrayItemInitialValues;
    }

    setFundingFormData({ index: itemIndex, values: fundingSourceValues });

    setOpenEditDialog(true);
  };

  const handleDialogEditSave = async (values: IProjectFundingFormArrayItem) => {
    const projectData = {
      funding: {
        fundingSources: [{ ...values }]
      }
    };

    const isEditing = fundingFormData.index < funding.fundingSources.length;
    const errorTitle = isEditing ? EditFundingI18N.editErrorTitle : AddFundingI18N.addErrorTitle;

    try {
      if (isEditing) {
        await restorationTrackerApi.project.updateProject(id, projectData);
      } else {
        await restorationTrackerApi.project.addFundingSource(id, projectData.funding.fundingSources[0]);
      }

      setOpenEditDialog(false);

      props.refresh();
    } catch (error) {
      const apiError = error as APIError;

      showErrorDialog({ dialogTitle: errorTitle, dialogText: apiError.message, open: true });
    }
  };

  const handleDeleteDialogOpen = async (itemIndex: number) => {
    setFundingFormData({
      index: itemIndex,
      values: funding.fundingSources[fundingFormData.index]
    });
    showYesNoDialog({ open: true });
  };

  const handleDeleteDialogYes = async () => {
    const fundingSource = funding.fundingSources[fundingFormData.index];

    try {
      await restorationTrackerApi.project.deleteFundingSource(id, fundingSource.id);
      showYesNoDialog({ open: false });
    } catch (error) {
      const apiError = error as APIError;
      showErrorDialog({
        dialogTitle: DeleteProjectFundingI18N.deleteErrorTitle,
        dialogText: apiError.message,
        open: true
      });
      return;
    }

    props.refresh();
  };

  const hasFundingSources = funding.fundingSources && funding.fundingSources.length > 0;

  return (
    <>
      <EditDialog
        dialogTitle={
          fundingFormData.index < funding.fundingSources.length ? EditFundingI18N.editTitle : AddFundingI18N.addTitle
        }
        open={openEditDialog}
        component={{
          element: (
            <ProjectFundingItemForm
              funding_sources={
                codes?.funding_source?.map((item) => {
                  return { value: item.id, label: item.name };
                }) || []
              }
              investment_action_category={
                codes?.investment_action_category?.map((item) => {
                  return { value: item.id, fs_id: item.fs_id, label: item.name };
                }) || []
              }
            />
          ),
          initialValues: fundingFormData.values,
          validationSchema: ProjectFundingFormArrayItemYupSchema
        }}
        onCancel={() => setOpenEditDialog(false)}
        onSave={handleDialogEditSave}
      />

      <H3ButtonToolbar
        label="Funding Sources"
        buttonLabel="Add Funding Source"
        buttonTitle="Add Funding Source"
        buttonStartIcon={<Icon path={mdiPlus} size={0.875} />}
        buttonOnClick={() => handleDialogEditOpen(funding.fundingSources.length)}
        toolbarProps={{ disableGutters: true }}
      />

      <TableContainer>
        <Table padding="default" className={classes.fundingSourceTable}>
          <TableHead>
            <TableRow>
              <TableCell>Agency</TableCell>
              <TableCell width="100px">Project ID</TableCell>
              <TableCell>Amount</TableCell>
              <TableCell>Dates</TableCell>
              <TableCell width="130px" align="center">
                Actions
              </TableCell>
            </TableRow>
          </TableHead>

          <TableBody>
            {hasFundingSources &&
              funding.fundingSources.map((item: any, index: number) => (
                <TableRow key={item.id}>
                  <TableCell>
                    {item.agency_name}
                    {item.investment_action_category_name !== 'Not Applicable' && (
                      <Typography component="em" variant="body2">
                        &nbsp;({item.investment_action_category_name})
                      </Typography>
                    )}
                  </TableCell>
                  <TableCell>
                    <Typography variant="body2" noWrap>
                      {item.agency_project_id || 'No Agency Project ID'}
                    </Typography>
                  </TableCell>
                  <TableCell>{getFormattedAmount(item.funding_amount)}</TableCell>
                  <TableCell>
                    <Typography variant="body2" noWrap>
                      {getFormattedDateRangeString(DATE_FORMAT.ShortMediumDateFormat, item.start_date, item.end_date)}
                    </Typography>
                  </TableCell>
                  <TableCell align="center">
                    <IconButton
                      onClick={() => handleDialogEditOpen(index)}
                      title="Edit Funding Source"
                      aria-label="Edit Funding Source"
                      data-testid="edit-funding-source">
                      <Icon path={mdiPencilOutline} size={0.875} />
                    </IconButton>
                    <IconButton
                      data-testid="delete-funding-source"
                      onClick={() => handleDeleteDialogOpen(index)}
                      title="Remove Funding Source"
                      aria-label="Remove Funding Source">
                      <Icon path={mdiTrashCanOutline} size={0.875} />
                    </IconButton>
                  </TableCell>
                </TableRow>
              ))}

            {!hasFundingSources && (
              <TableRow>
                <TableCell colSpan={5}>No Funding Sources</TableCell>
              </TableRow>
            )}
          </TableBody>
        </Table>
      </TableContainer>
    </>
  );
};

export default FundingSource;
