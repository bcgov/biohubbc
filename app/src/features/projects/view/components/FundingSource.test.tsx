import { fireEvent, render, waitFor, cleanup, within } from '@testing-library/react';
import { DialogContextProvider } from 'contexts/dialogContext';
import { useRestorationTrackerApi } from 'hooks/useRestorationTrackerApi';
import React from 'react';
import { codes } from 'test-helpers/code-helpers';
import { getProjectForViewResponse } from 'test-helpers/project-helpers';
import FundingSource from './FundingSource';
jest.mock('../../../../hooks/useRestorationTrackerApi');
const mockRefresh = jest.fn();

jest.mock('../../../../hooks/useRestorationTrackerApi');
const mockuseRestorationTrackerApi = {
  project: {
    updateProject: jest.fn(),
    addFundingSource: jest.fn(),
    deleteFundingSource: jest.fn()
  }
};

const mockRestorationTrackerApi = ((useRestorationTrackerApi as unknown) as jest.Mock<
  typeof mockuseRestorationTrackerApi
>).mockReturnValue(mockuseRestorationTrackerApi);

describe('FundingSource', () => {
  beforeEach(() => {
    // clear mocks before each test
    mockRestorationTrackerApi().project.updateProject.mockClear();
    mockRestorationTrackerApi().project.addFundingSource.mockClear();
    mockRestorationTrackerApi().project.deleteFundingSource.mockClear();
  });

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const { asFragment } = render(
      <FundingSource projectForViewData={getProjectForViewResponse} codes={codes} refresh={mockRefresh} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('opens the edit funding source dialog box when edit button is clicked, and cancel button works as expected', async () => {
    const { getByText, getByTestId, queryByText } = render(
      <FundingSource projectForViewData={getProjectForViewResponse} codes={codes} refresh={mockRefresh} />
    );

    await waitFor(() => {
      expect(getByText('Funding Sources')).toBeInTheDocument();
    });

    fireEvent.click(getByTestId('edit-funding-source'));

    await waitFor(() => {
      expect(getByText('Edit Funding Source')).toBeVisible();
    });

    fireEvent.click(getByText('Cancel'));

    await waitFor(() => {
      expect(queryByText('Edit Funding Source')).not.toBeInTheDocument();
    });
  });

  it('edits a funding source correctly in the dialog', async () => {
    const { getByText, getByTestId } = render(
      <FundingSource projectForViewData={getProjectForViewResponse} codes={codes} refresh={mockRefresh} />
    );

    await waitFor(() => {
      expect(getByText('Funding Sources')).toBeInTheDocument();
    });

    fireEvent.click(getByTestId('edit-funding-source'));

    await waitFor(() => {
      expect(getByText('Agency Details')).toBeVisible();
    });

    fireEvent.click(getByText('Save Changes'));

    await waitFor(() => {
      expect(mockRestorationTrackerApi().project.updateProject).toHaveBeenCalledTimes(1);
      expect(mockRefresh).toBeCalledTimes(1);
    });
  });

  it('shows error dialog with API error message when editing a funding source fails', async () => {
    mockRestorationTrackerApi().project.updateProject = jest.fn(() => Promise.reject(new Error('API Error is Here')));

    const { getByText, getByTestId, queryByText, getAllByRole } = render(
      <DialogContextProvider>
        <FundingSource projectForViewData={getProjectForViewResponse} codes={codes} refresh={mockRefresh} />
      </DialogContextProvider>
    );

    await waitFor(() => {
      expect(getByText('Funding Sources')).toBeInTheDocument();
    });

    fireEvent.click(getByTestId('edit-funding-source'));

    await waitFor(() => {
      expect(getByText('Agency Details')).toBeVisible();
    });

    fireEvent.click(getByText('Save Changes'));

    await waitFor(() => {
      expect(queryByText('API Error is Here')).toBeInTheDocument();
    });

    // Get the backdrop, then get the firstChild because this is where the event listener is attached
    //@ts-ignore
    fireEvent.click(getAllByRole('presentation')[0].firstChild);

    await waitFor(() => {
      expect(queryByText('API Error is Here')).toBeNull();
    });
  });

  it('deletes a funding source as expected', async () => {
    const { getByText, getByTestId } = render(
      <DialogContextProvider>
        <FundingSource projectForViewData={getProjectForViewResponse} codes={codes} refresh={mockRefresh} />
      </DialogContextProvider>
    );

    await waitFor(() => {
      expect(getByText('Funding Sources')).toBeInTheDocument();
    });

    fireEvent.click(getByTestId('delete-funding-source'));

    await waitFor(() => {
      expect(
        getByText(
          'Are you sure you want to remove this project funding source? It will also remove the associated survey funding source.'
        )
      ).toBeVisible();
    });

    fireEvent.click(getByText('Yes'));

    await waitFor(() => {
      expect(mockRestorationTrackerApi().project.deleteFundingSource).toHaveBeenCalledTimes(1);
      expect(mockRefresh).toBeCalledTimes(1);
    });
  });

  it('closes the delete dialog when user decides not to delete their funding source', async () => {
    const { getByText, queryByText, getByTestId, getAllByRole } = render(
      <DialogContextProvider>
        <FundingSource projectForViewData={getProjectForViewResponse} codes={codes} refresh={mockRefresh} />
      </DialogContextProvider>
    );

    await waitFor(() => {
      expect(getByText('Funding Sources')).toBeInTheDocument();
    });

    fireEvent.click(getByTestId('delete-funding-source'));

    await waitFor(() => {
      expect(
        getByText(
          'Are you sure you want to remove this project funding source? It will also remove the associated survey funding source.'
        )
      ).toBeVisible();
    });

    fireEvent.click(getByText('No'));

    await waitFor(() => {
      expect(
        queryByText(
          'Are you sure you want to remove this project funding source? It will also remove the associated survey funding source.'
        )
      ).toBeNull();
    });

    fireEvent.click(getByTestId('delete-funding-source'));

    await waitFor(() => {
      expect(
        getByText(
          'Are you sure you want to remove this project funding source? It will also remove the associated survey funding source.'
        )
      ).toBeVisible();
    });

    // Get the backdrop, then get the firstChild because this is where the event listener is attached
    //@ts-ignore
    fireEvent.click(getAllByRole('presentation')[0].firstChild);

    await waitFor(() => {
      expect(
        queryByText(
          'Are you sure you want to remove this project funding source? It will also remove the associated survey funding source.'
        )
      ).toBeNull();
    });
  });

  it('shows error dialog with API error message when deleting a funding source fails', async () => {
    mockRestorationTrackerApi().project.deleteFundingSource = jest.fn(() =>
      Promise.reject(new Error('API Error is Here'))
    );

    const { getByText, queryByText, getByTestId } = render(
      <DialogContextProvider>
        <FundingSource projectForViewData={getProjectForViewResponse} codes={codes} refresh={mockRefresh} />
      </DialogContextProvider>
    );

    await waitFor(() => {
      expect(getByText('Funding Sources')).toBeInTheDocument();
    });

    fireEvent.click(getByTestId('delete-funding-source'));

    await waitFor(() => {
      expect(
        getByText(
          'Are you sure you want to remove this project funding source? It will also remove the associated survey funding source.'
        )
      ).toBeVisible();
    });

    fireEvent.click(getByText('Yes'));

    await waitFor(() => {
      expect(queryByText('API Error is Here')).toBeInTheDocument();
    });

    fireEvent.click(getByText('Ok'));

    await waitFor(() => {
      expect(queryByText('API Error is Here')).toBeNull();
    });
  });

  it('adds a funding source as expected', async () => {
    const { getByText, getByTestId, getAllByRole, getByRole } = render(
      <FundingSource
        projectForViewData={{
          ...getProjectForViewResponse,
          funding: { ...getProjectForViewResponse.funding, fundingSources: [] }
        }}
        codes={codes}
        refresh={mockRefresh}
      />
    );

    await waitFor(() => {
      expect(getByText('Funding Sources')).toBeInTheDocument();
    });

    fireEvent.click(getByText('Add Funding Source'));

    await waitFor(() => {
      expect(getByText('Agency Details')).toBeInTheDocument();
    });

    /*
      Triggering onChange on Material UI Select elements
      https://stackoverflow.com/questions/55184037/react-testing-library-on-change-for-material-ui-select-component
    */
    fireEvent.mouseDown(getAllByRole('button')[0]);
    const agencyNameListbox = within(getByRole('listbox'));
    fireEvent.click(agencyNameListbox.getByText(/Funding source code/i));

    await waitFor(() => {
      expect(getByTestId('investment_action_category')).toBeInTheDocument();
    });

    fireEvent.mouseDown(getAllByRole('button')[1]);

    const investmentActionCategoryListbox = within(getByRole('listbox'));

    fireEvent.click(investmentActionCategoryListbox.getByText(/Investment action category/i));
    fireEvent.change(getByTestId('funding_amount'), { target: { value: 100 } });
    fireEvent.change(getByTestId('start-date'), { target: { value: '2021-03-14' } });
    fireEvent.change(getByTestId('end-date'), { target: { value: '2021-05-14' } });

    fireEvent.click(getByText('Save Changes'));

    await waitFor(() => {
      expect(mockRestorationTrackerApi().project.addFundingSource).toHaveBeenCalledTimes(1);
      expect(mockRefresh).toBeCalledTimes(1);
    });
  });
});
