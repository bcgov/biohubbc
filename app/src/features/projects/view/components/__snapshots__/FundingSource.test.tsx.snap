// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FundingSource renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiToolbar-root MuiToolbar-regular"
    style="justify-content: space-between;"
  >
    <h3
      class="MuiTypography-root MuiTypography-h3 MuiTypography-colorInherit"
    >
      Funding Sources
    </h3>
    <div
      class="MuiBox-root MuiBox-root-2"
    >
      <button
        aria-label="Add Funding Source"
        class="MuiButtonBase-root MuiButton-root MuiButton-text sectionHeaderButton MuiButton-textPrimary"
        data-testid="h3-button-toolbar-AddFundingSource"
        id="h3-button-toolbar-AddFundingSource"
        tabindex="0"
        title="Add Funding Source"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              role="presentation"
              style="width: 1.3125rem; height: 1.3125rem;"
              viewBox="0 0 24 24"
            >
              <path
                d="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
                style="fill: currentColor;"
              />
            </svg>
          </span>
          <strong>
            Add Funding Source
          </strong>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiTableContainer-root"
  >
    <table
      class="MuiTable-root makeStyles-fundingSourceTable-1"
    >
      <thead
        class="MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            Agency
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head"
            scope="col"
            width="100px"
          >
            Project ID
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            Amount
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            Dates
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
            width="130px"
          >
            Actions
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            agency name
            <em
              class="MuiTypography-root MuiTypography-body2"
            >
              Â (investment action)
            </em>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap"
            >
              ABC123
            </p>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            $333
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap"
            >
              Apr 14, 2000 - Apr 13, 2021
            </p>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <button
              aria-label="Edit Funding Source"
              class="MuiButtonBase-root MuiIconButton-root"
              data-testid="edit-funding-source"
              tabindex="0"
              title="Edit Funding Source"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  role="presentation"
                  style="width: 1.3125rem; height: 1.3125rem;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M14.06,9L15,9.94L5.92,19H5V18.08L14.06,9M17.66,3C17.41,3 17.15,3.1 16.96,3.29L15.13,5.12L18.88,8.87L20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18.17,3.09 17.92,3 17.66,3M14.06,6.19L3,17.25V21H6.75L17.81,9.94L14.06,6.19Z"
                    style="fill: currentColor;"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="Remove Funding Source"
              class="MuiButtonBase-root MuiIconButton-root"
              data-testid="delete-funding-source"
              tabindex="0"
              title="Remove Funding Source"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  role="presentation"
                  style="width: 1.3125rem; height: 1.3125rem;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M9,3V4H4V6H5V19A2,2 0 0,0 7,21H17A2,2 0 0,0 19,19V6H20V4H15V3H9M7,6H17V19H7V6M9,8V17H11V8H9M13,8V17H15V8H13Z"
                    style="fill: currentColor;"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
